proposals/nnnn-strict-memory-safety.md

```diff
@@ -0,0 +1,536 @@
+
+### Unsafe conformances
+
+Implementing a protocol requirement that is safe (and not part of an `@unsafe` protocol) within an `@unsafe` declaration introduces unsafety, so it will produce a diagnostic in the strict safety mode:
+
+```swift
+protocol P {
+  func f()
+}
+
+struct ConformsToP { }
+
+extension ConformsToP: P {
+  @unsafe func f() { } // warning: unsafe instance method 'f()' cannot satisfy safe requirement
+}
+```
+
+To suppress this warning, one can place `@unsafe` on the conformance to `P` is supplied. This notes that the conformance itself is unsafe:
+
+```swift
+extension ConformsToP: @unsafe P {
+  @unsafe func f() { }
+}
+```
+
+Use of an `@unsafe` conformance for any reason (e.g., when that conformance is needed to call a generic function with a `ConformsToP` requirement) is diagnosed as an unsafe use, much like use of an `@unsafe` declaration. For example
+
+```swift
+func acceptP<T: P>(_: T.Type) { }
+
+func passUnsafe() {
+  acceptP(ConformsToP.self) // warning: use of @unsafe conformance of 'ConformsToP' to protocol 'P'
+}
+```
+
+### Strict safety mode and escalatable warnings
+
+The strict memory safety mode can be enabled with the new compiler flag `-strict-memory-safety`.
+
+All of the memory-safety diagnostics produced by the strict memory safety mode will be warnings. These warnings be in the group `Unsafe` (possibly organized into subgroups) so that one can choose to escalate them to errors or keep them as warnings using the compiler flags introduced in [SE-0443](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0443-warning-control-flags.md). For example, one can choose to enable the mode and make memory-safety issues errors using:
+
+```
+swiftc -strict-memory-safety -Werror Unsafe
+```
+
+### SwiftPM integration
+
+Swift package manifests will need a way to enable strict memory safety mode on a per-module and per-package basis. This proposal extends the `SwiftSetting` type in the manifest with a new option to enable that checking:
+
+```swift
+static func strictMemorySafety(
+  _ condition: BuildSettingCondition? = nil
+) -> SwiftSetting
+```
+
+## Source compatibility
+
+The `unsafe` keyword in this proposal will be introduced as a contextual keyword following the precedent set by `await`'s' introduction in [SE-0296](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0296-async-await.md) and `consume`'s introduction in [SE-0366](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0366-move-function.md). This allows `unsafe` to continue to be used as an identifier, albeit with a small potential to break existing source that uses `unsafe` as a function that is then called with a trailing closure, like this:
+
+```swift
+func unsafe(_ body: () -> Void) { }
+
+unsafe {
+  // currently calls 'unsafe(_:)', but will become an unsafe expression
+}
```

Comment:
Interesting. This is still necessary with the removal of `unsafe` blocks? Because this expression seems to be a no-op. I guess having a different interpretation rule for `unsafe { ... }` would be a little weird, but maybe we can at least warn that it doesn't cover any code that could possibly be unsafe.

Reply:
We do warn if it doesn't cover any code that's unsafe. If the closure had some unsafe types in its signature, then that warning wouldn't trigger.